processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -149,4 +155,178 @@ public void remove()
+        List<Filter> children = Lists.newArrayList();
+        for (Filter grandChild : ((AndFilter) child).getFilters()) {
+          children.add(pushDownNot(new NotFilter(grandChild)));
+        }
+        return new OrFilter(children);
+      }
+      if (child instanceof OrFilter) {
+        List<Filter> children = Lists.newArrayList();
+        for (Filter grandChild : ((OrFilter) child).getFilters()) {
+          children.add(pushDownNot(new NotFilter(grandChild)));
+        }
+        return new AndFilter(children);
+      }
+    }
+
+
+    if (current instanceof AndFilter) {
+      List<Filter> children = Lists.newArrayList();
+      for (Filter child : ((AndFilter) current).getFilters()) {
+        children.add(pushDownNot(child));
+      }
+      return new AndFilter(children);
+    }
+
+
+
+    if (current instanceof OrFilter) {
+      List<Filter> children = Lists.newArrayList();
+      for (Filter child : ((OrFilter) current).getFilters()) {
+        children.add(pushDownNot(child));
+      }
+      return new OrFilter(children);
+    }
+    return current;
+  }
+
+  // CNF conversion functions were adapted from Apache Hive, see:
+  // https://github.com/apache/hive/blob/branch-2.0/storage-api/src/java/org/apache/hadoop/hive/ql/io/sarg/SearchArgumentImpl.java
+  private static Filter convertToCNFInternal(Filter current)
+  {
+    if (current instanceof NotFilter) {
+      return new NotFilter(convertToCNFInternal(((NotFilter) current).getBaseFilter()));
+    }
+    if (current instanceof AndFilter) {
+      List<Filter> children = Lists.newArrayList();
+      for (Filter child : ((AndFilter) current).getFilters()) {
+        children.add(convertToCNFInternal(child));
+      }
+      return new AndFilter(children);
+    }
+    if (current instanceof OrFilter) {
+      // a list of leaves that weren't under AND expressions
+      List<Filter> nonAndList = new ArrayList<Filter>();
+      // a list of AND expressions that we need to distribute
+      List<Filter> andList = new ArrayList<Filter>();
+      for (Filter child : ((OrFilter) current).getFilters()) {
+        if (child instanceof AndFilter) {
+          andList.add(child);
+        } else if (child instanceof OrFilter) {
+          // pull apart the kids of the OR expression
+          for (Filter grandChild : ((OrFilter) child).getFilters()) {
+            nonAndList.add(grandChild);
+          }
+        } else {
+          nonAndList.add(child);
+        }
+      }
+      if (!andList.isEmpty()) {
+        List<Filter> result = Lists.newArrayList();
+        generateAllCombinations(result, andList, nonAndList);
+        return new AndFilter(result);
+      }
+    }
+    return current;
+  }
+
+  // CNF conversion functions were adapted from Apache Hive, see:
+  // https://github.com/apache/hive/blob/branch-2.0/storage-api/src/java/org/apache/hadoop/hive/ql/io/sarg/SearchArgumentImpl.java
+  private static Filter flatten(Filter root) {
```

Comment:
This method is kind of awkward… it's modifying Filter objects in-place, I think we'd generally prefer to not to that.